structure/tree/rbtree.go

```diff
@@ -0,0 +1,352 @@
+//
+// For more details check out those links below here:
+// Programiz article : https://www.programiz.com/dsa/red-black-tree
+// Wikipedia article: https://en.wikipedia.org/wiki/Red_black_tree
+// authors [guuzaa](https://github.com/guuzaa)
+// see rbtree.go
+
+package tree
+
+import (
+	"fmt"
+
+	"github.com/TheAlgorithms/Go/constraints"
+)
+
+type RBTree[T constraints.Ordered] struct {
+	*binaryTree[T]
+}
+
+// Create a new Red-Black Tree
+func NewRBTree[T constraints.Ordered]() *RBTree[T] {
+	leaf := &Node[T]{Color: Black, Left: nil, Right: nil}
+	return &RBTree[T]{
+		binaryTree: &binaryTree[T]{
+			Root: leaf,
+			NIL:  leaf,
+		},
+	}
+}
+
+// Insert a chain of Node's into the Red-Black Tree
+func (t *RBTree[T]) Insert(keys ...T) {
+	for _, key := range keys {
+		t.insertHelper(t.Root, key)
+	}
+}
+
+// Delete a node of Red-Black Tree
+// Returns false if the node does not exist, otherwise returns true.
+func (t *RBTree[T]) Delete(data T) bool {
+	return t.deleteHelper(t.Root, data)
+}
+
+// Return the Predecessor of the node of Key
+// if there is no predecessor, return default value of type T and false
+// otherwise return the Key of predecessor and true
+func (t *RBTree[T]) Predecessor(key T) (T, bool) {
+	node, ok := t.searchTreeHelper(t.Root, key)
+	if !ok {
+		return t.NIL.Key, ok
+	}
+	return t.predecessorHelper(node)
+}
+
+// Return the Successor of the node of Key
+// if there is no successor, return default value of type T and false
+// otherwise return the Key of successor and true
+func (t *RBTree[T]) Successor(key T) (T, bool) {
+	node, ok := t.searchTreeHelper(t.Root, key)
+	if !ok {
+		return t.NIL.Key, ok
+	}
+
+	return t.successorHelper(node)
+}
+
+func (t *RBTree[T]) insertHelper(x *Node[T], key T) {
+	node := &Node[T]{
+		Key:    key,
+		Left:   t.NIL,
+		Right:  t.NIL,
+		Parent: nil,
+		Color:  Red,
+	}
+
+	var y *Node[T]
+	for !t.isNil(x) {
+		y = x
+		if node.Key < x.Key {
+			x = x.Left
+		} else {
+			x = x.Right
+		}
+
+	}
+
+	node.Parent = y
+	if y == nil {
+		t.Root = node
+	} else if node.Key < y.Key {
+		y.Left = node
+	} else {
+		y.Right = node
+	}
+
+	if node.Parent == nil {
+		node.Color = Black
+		return
+	}
+
+	if node.Parent.Parent == nil {
+		return
+	}
+
+	t.insertFix(node)
+}
+
+func (t *RBTree[T]) leftRotate(x *Node[T]) {
+	y := x.Right
+	x.Right = y.Left
+
+	if !t.isNil(y.Left) {
+		y.Left.Parent = x
+	}
+
+	y.Parent = x.Parent
+	if x.Parent == nil {
+		t.Root = y
+	} else if x == x.Parent.Left {
+		x.Parent.Left = y
+	} else {
+		x.Parent.Right = y
+	}
+
+	y.Left = x
+	x.Parent = y
+}
+
+func (t *RBTree[T]) rightRotate(x *Node[T]) {
+	y := x.Left
+	x.Left = y.Right
+	if !t.isNil(y.Right) {
+		y.Right.Parent = x
+	}
+	y.Parent = x.Parent
+	if x.Parent == nil {
+		t.Root = y
+	} else if x == y.Parent.Right {
+		y.Parent.Right = y
+	} else {
+		y.Parent.Left = y
+	}
+
+	y.Right = x
+	x.Parent = y
+}
+
+func (t *RBTree[T]) insertFix(k *Node[T]) {
```

Comment:
Can you think of a way to reduce the nesting? I think for beginners it's quite difficult to follow this...

Reply:
I will try to do it. Honestly, I think it's impossible. As we all known, Red-Black tree is complex. 😭